### ISSUE TITLE
Custom post type lock breaks block recovery

### ISSUE THREAD

**Issue by Tropicalista** [WordPress/gutenberg#49005]
### Description

Adding a template lock can break the block recovery.

This line prevent block recovery: https://github.com/WordPress/gutenberg/blob/605aeb0f4f7d2225120e498f95ae27b9f56d77a3/packages/block-editor/src/store/selectors.js#L1538

### Step-by-step reproduction instructions

1. Create a custom post type with a template `core/group`
2. Add `template_lock: 'insert'`
3. Create new custom post type and add a paragraph or whatever
4. In the code editor change markup of `core/group` addin something like `data-foo="bar"`
5. The block is invalid. 
6. Try to recovery block and nothing happens

Do the same steps on a normal post/page. The recovery works great if you try the recovery on a post/page.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by Tropicalista** [comment]
The expected behaviour is to recovery the block. However this is not possible due to a lock.
But recovery does not insert new block. It just replace the existing one with same attributes. The lock check should be bypassed in this case.

**Comment by Mamaduka** [comment]
I can reproduce the issue.

When a template is locked, we should probably do a more granular check inside `canInsertBlockTypeUnmemoized`. Maybe something similar to `validateBlocksToTemplate`.

**Comment by talldan** [comment]
> When a template is locked, we should probably do a more granular check inside canInsertBlockTypeUnmemoized. Maybe something similar to validateBlocksToTemplate.

Yeah, it's a difficult one to solve. `canInsertBlockType` has no knowledge of the block being replaced, it just knows that the parent block is locked and so we can't insert the recovered block.

Maybe introducing a `canReplaceBlock` selector that does a slightly different check and accounts for block validity, would be the way forwards.

An observation - I notice that when an invalid block has 'prevent removal' active, it can still be recovered 🤔

**Comment by Tropicalista** [comment]
@talldan exactly what I was thinking: introduce a canReplaceBlock can also be used to insert pattern.

For example if pattern selected match root block, it will be replaced, otherwise skipped.

Another solution can be to introduce a block limit count on post type that will act as a locker.

**Comment by youknowriad** [comment]
Still reproduces on trunk (recovery does nothing under a template lock, works without one). This is the same root cause as #25216, so closing as a duplicate to keep it in one place.
